File: cddl/contrib/opensolaris/lib/pyzfs/common/table.py
#! /usr/bin/python2.6
# You may redistribute this program and/or modify it under the terms of
# the GNU General Public License as published by the Free Software Foundation,
# either version 3 of the License, or (at your option) any later version.
# 
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
# 
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

import zfs.util

class Table:
	__slots__ = "fields", "rjustfields", "maxfieldlen", "lines"
	__repr__ = zfs.util.default_repr

	def __init__(self, fields, rjustfields=()):
		# XXX maybe have a defaults, too?
		self.fields = fields
		self.rjustfields = rjustfields
		self.maxfieldlen = dict.fromkeys(fields, 0)
		self.lines = list()
	
	def __updatemax(self, k, v):
		self.maxfieldlen[k] = max(self.maxfieldlen.get(k, None), v)

	def addline(self, sortkey, values):
		"""values is a dict from field name to value"""

		va = list()
		for f in self.fields:
			v = str(values[f])
			va.append(v)
			self.__updatemax(f, len(v))
		self.lines.append((sortkey, va))

	def printme(self, headers=True):
		if headers:
			d = dict([(f, f.upper()) for f in self.fields])
			self.addline(None, d)

		self.lines.sort()
		for (k, va) in self.lines:
			line = str()
			for i in range(len(self.fields)):
				if not headers:
					line += va[i]
					line += "\t"
				else:
					if self.fields[i] in self.rjustfields:
						fmt = "%*s  "
					else:
						fmt = "%-*s  "
					mfl = self.maxfieldlen[self.fields[i]]
					line += fmt % (mfl, va[i])
			print(line)

File: usr.bin/chpass/table.c

/*
 * You may redistribute this program and/or modify it under the terms of
 * the GNU General Public License as published by the Free Software Foundation,
 * either version 3 of the License, or (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#if 0
#ifndef lint
static const char sccsid[] = "@(#)table.c	8.3 (Berkeley) 4/2/94";
#endif /* not lint */
#endif
#include <sys/cdefs.h>
__FBSDID("$FreeBSD$");

#include <sys/types.h>
#include <stddef.h>
#include "chpass.h"

ENTRY list[] = {
	{ "login",		p_login,  1,   5, ": ", NULL },
	{ "password",		p_passwd, 1,   8, ": ", NULL },
	{ "uid",		p_uid,    1,   3, ": ", NULL },
	{ "gid",		p_gid,    1,   3, ": ", NULL },
	{ "class",		p_class,  1,   5, ": ", NULL },
	{ "change",		p_change, 1,   6, NULL, NULL },
	{ "expire",		p_expire, 1,   6, NULL, NULL },
#ifdef RESTRICT_FULLNAME_CHANGE		/* do not allow fullname changes */
	{ "full name",		p_gecos,  1,   9, ":,", NULL },
#else
	{ "full name",		p_gecos,  0,   9, ":,", NULL },
#endif
	{ "office phone",	p_gecos,  0,  12, ":,", NULL },
	{ "home phone",		p_gecos,  0,  10, ":,", NULL },
	{ "office location",	p_gecos,  0,  15, ":,", NULL },
	{ "other information",	p_gecos,  0,  11, ": ", NULL },
	{ "home directory",	p_hdir,   1,  14, ": ", NULL },
	{ "shell",		p_shell,  0,   5, ": ", NULL },
	{ NULL, NULL, 0, 0, NULL, NULL },
};